### PR TITLE
feat(tickets): add menu item selection for free premium perk

### LINKED ISSUE #582: Selecting menu items on instant purchases
It should be required to choose menu item when redeeming a free purchase such as perks, contributing to the statistics of which products are sold.

When selecting a one-ticket purchase for a product with multiple items, a dropdown should appear where the selection of menu items is available. This selection should be passed on to the receipt shown after a swipe/purchase for the barista in the bar to see.

### PR DESCRIPTION
**Pull request by duckth**
Closes #582 
This introduces a drop down on free perks with multiple menu items, allowing the user to select which item they are claiming. This is to give better tracking of the menu items being "purchased", even with premium perks, and to lessen confusion by baristas as the default receipt just says "Espresso".

### CODE REVIEW

**Review by StineHelena — APPROVED** [review]
Looks good, nice job :-)

**Comment by marfavi** [comment]
Made it prettier + made it remember which menu item you swiped last time :)

It required refactoring: the way the app used to remember which menu item you swiped was sitting on the Ticket class. Now, we put the functionality on Product instead, which allows both ticket swipes + perk claims to use the feature.

File: lib/features/product/presentation/functions.dart [hunk 1]
```diff
@@ -70,6 +70,7 @@ Future<void> _afterPurchaseModal(
   if (singleTicketPurchase) {
     await ticketsCubit.useTicket(
       product.id,
+      // The first eligible menu item is used as the default.
```
Comment (marfavi) [inline lib/features/product/presentation/functions.dart:73]: We don't have this functionality as part of this PR, but I added a comment acknowledging this

File: lib/features/product/presentation/widgets/buy_ticket_bottom_modal_sheet.dart [hunk 2]
```diff
@@ -96,34 +96,6 @@ class _BottomModalSheetButtonBarState
   Widget build(BuildContext context) {
     final productPrice = widget.product.price;
     final productId = widget.product.id;
-    final isFreeProduct = productPrice == 0;
```
Comment (marfavi) [inline lib/features/product/presentation/widgets/buy_ticket_bottom_modal_sheet.dart:99]: This doesn't show for free products anymore

File: lib/features/product/product_model.dart [hunk 3]
```diff
@@ -52,6 +54,31 @@ class Product extends Equatable {
   final bool isPerk;
   final List<MenuItem> eligibleMenuItems;
 
+  /// The name of the Hive box used to store the last used menu item.
+  static const _hiveBoxName = 'lastUsedMenuItemByProductId';
+
+  /// Returns the last used menu item for this product as a [TaskOption].
+  TaskOption<MenuItem> get lastUsedMenuItem {
```
Comment (marfavi) [inline lib/features/product/product_model.dart:61]: moved from Ticket class

File: lib/features/ticket/domain/entities/ticket.dart [hunk 4]
```diff
@@ -1,36 +1,26 @@
-import 'package:coffeecard/features/product/menu_item_model.dart';
```
Comment (marfavi) [inline lib/features/ticket/domain/entities/ticket.dart:1]: class has been simplified now that Product manages the last used menu item

File: lib/features/ticket/domain/usecases/load_tickets.dart [hunk 5]
```diff
@@ -35,46 +33,11 @@ class LoadTickets {
...
           (product) => Ticket(
-            product: product,
             amountLeft: tickets.length,
-            lastUsedMenuItem: const None(),
+            product: product,
           ),
-        )
-        .flatMap(_addLastUsedMenuItemToTicket);
-  }
-
```
Comment (marfavi) [inline lib/features/ticket/domain/usecases/load_tickets.dart:49]: moved to Product

File: lib/features/ticket/presentation/widgets/perk_card.dart [hunk 6]
```diff
@@ -22,18 +24,50 @@ class PerkCard extends StatelessWidget {
...
-      icon: Icons.coffee_maker_outlined,
-      onTapped: (_) => buyModal(
+  void _onTapped(BuildContext context) {
+    if (product.price == 0 && product.amount == 1) {
+      showClaimSinglePerkConfirm(
+        context: context,
+        product: product,
+      );
+    } else if (product.price == 0 && product.amount > 1) {
+      appDialog(
```
Comment (marfavi) [inline lib/features/ticket/presentation/widgets/perk_card.dart:34]: Shouldn't happen in prod, but claiming > 1 free product at a time is not intended behaviour. We show an error in this case

**Review by marfavi — APPROVED** [review]
Approved after joint review in person

**Pull request merged**